src/cpp/common/alarm.cc

```diff
@@ -52,78 +62,94 @@ class AlarmImpl : public grpc::internal::CompletionQueueTag {
     return true;
   }
   void Set(grpc::CompletionQueue* cq, gpr_timespec deadline, void* tag) {
-    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
     grpc_core::ExecCtx exec_ctx;
     GRPC_CQ_INTERNAL_REF(cq->cq(), "alarm");
     cq_ = cq->cq();
     tag_ = tag;
     GPR_ASSERT(grpc_cq_begin_op(cq_, this));
-    GRPC_CLOSURE_INIT(
-        &on_alarm_,
-        [](void* arg, grpc_error_handle error) {
-          // queue the op on the completion queue
-          AlarmImpl* alarm = static_cast<AlarmImpl*>(arg);
-          alarm->Ref();
-          // Preserve the cq and reset the cq_ so that the alarm
-          // can be reset when the alarm tag is delivered.
-          grpc_completion_queue* cq = alarm->cq_;
-          alarm->cq_ = nullptr;
-          grpc_cq_end_op(
-              cq, alarm, error,
-              [](void* /*arg*/, grpc_cq_completion* /*completion*/) {}, arg,
-              &alarm->completion_);
-          GRPC_CQ_INTERNAL_UNREF(cq, "alarm");
-        },
-        this, grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&timer_,
-                    grpc_core::Timestamp::FromTimespecRoundUp(deadline),
-                    &on_alarm_);
+    Ref();
+    GPR_ASSERT(cq_armed_.exchange(true) == false);
+    GPR_ASSERT(!callback_armed_.load());
+    cq_timer_handle_ = event_engine_->RunAfter(
+        grpc_core::Timestamp::FromTimespecRoundUp(deadline) -
+            grpc_core::ExecCtx::Get()->Now(),
+        [this] {
+          cq_armed_.store(false);
+          OnCQAlarm(absl::OkStatus());
+        });
   }
   void Set(gpr_timespec deadline, std::function<void(bool)> f) {
-    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
     grpc_core::ExecCtx exec_ctx;
     // Don't use any CQ at all. Instead just use the timer to fire the function
     callback_ = std::move(f);
     Ref();
-    GRPC_CLOSURE_INIT(
-        &on_alarm_,
-        [](void* arg, grpc_error_handle error) {
-          grpc_core::Executor::Run(GRPC_CLOSURE_CREATE(
-                                       [](void* arg, grpc_error_handle error) {
-                                         AlarmImpl* alarm =
-                                             static_cast<AlarmImpl*>(arg);
-                                         alarm->callback_(error.ok());
-                                         alarm->Unref();
-                                       },
-                                       arg, nullptr),
-                                   error);
-        },
-        this, grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&timer_,
-                    grpc_core::Timestamp::FromTimespecRoundUp(deadline),
-                    &on_alarm_);
+    GPR_ASSERT(callback_armed_.exchange(true) == false);
+    GPR_ASSERT(!cq_armed_.load());
```

Comment:
So an alarm can be either `callback_armed_` or `cq_armed_`, but not both? Is this semantic documented somewhere? What will happen if we allow both?

Reply:
Oh, just saw the original code use only one set of `grpc_timer timer_;` and `grpc_closure on_alarm_;`, nvm then.

Reply:
I added documentation to the public API. The multi-set capability wasn't documented either, but we do test for it, so I added a note on that as well.

> What will happen if we allow both?

In the previous implementation, the timer_ and on_alarm_ objects were shared between the CQ and Callback implementations. They would have been overwritten (or some other UB, I think, since the iomgr timer API is not clear on re-use) if both `Set` methods were used concurrently.

Reply:
In this PR, if both are used concurrently, an assert will fail. I prefer this to UB (likely missing events).